File: src-tauri/src/commands/process.rs
// 进程控制命令 — 使用 libloading 动态加载 ntdll Native API
// Process control commands — uses libloading to dynamically load ntdll Native API
use crate::services::process_control_service::{
    resume_process_by_pid, suspend_process_by_pid, terminate_process_by_pid,
};
use crate::services::process_monitor_service::ProcessMonitorService;
use tauri::Manager;

/// 函数名称：suspend_process
/// 函数作用：挂起指定 PID 的进程（通过 NtSuspendProcess 动态调用）。
/// Purpose: Suspends the process identified by PID (via dynamic NtSuspendProcess call).
/// 中文关键词：挂起进程，暂停进程，进程控制，NtSuspendProcess
/// English keywords: suspend process, pause process, process control, NtSuspendProcess
#[tauri::command]
pub async fn suspend_process(pid: u32) -> Result<bool, String> {
    suspend_process_by_pid(pid)
}

/// 函数名称：resume_process
/// 函数作用：恢复指定 PID 的进程（通过 NtResumeProcess 动态调用）。
/// Purpose: Resumes the process identified by PID (via dynamic NtResumeProcess call).
/// 中文关键词：恢复进程，继续进程，进程控制，NtResumeProcess
/// English keywords: resume process, continue process, process control, NtResumeProcess
#[tauri::command]
pub async fn resume_process(pid: u32) -> Result<bool, String> {
    resume_process_by_pid(pid)
}

/// 函数名称：terminate_process
/// 函数作用：终止指定 PID 的进程。
/// Purpose: Terminates the process identified by PID.
/// 中文关键词：终止进程，结束进程，进程控制，TerminateProcess
/// English keywords: terminate process, kill process, process control, TerminateProcess
#[tauri::command]
pub async fn terminate_process(pid: u32) -> Result<bool, String> {
    terminate_process_by_pid(pid)
}

/// 函数名称：start_process_watcher
/// 函数作用：启动进程监控服务，轮询新进程并对未签名进程注入 file_hook；路径参数为空时由后端解析默认开发目录或打包资源路径。
/// Purpose: Starts the process monitor service, polls new processes and injects file_hook into unsigned ones; empty path parameters are resolved from development or bundled resource defaults.
/// 调用方：前端设置页面
/// Called by: Frontend settings page
/// 被调用方：ProcessMonitorService::start_with_resource_dir，AppHandle::path().resource_dir。
/// Calls: ProcessMonitorService::start_with_resource_dir, AppHandle::path().resource_dir.
/// 参数说明：injector_x64/injector_x86/dll_x64/dll_x86 为空时使用后端默认路径；interval_ms 为轮询间隔。
/// Parameters: injector_x64/injector_x86/dll_x64/dll_x86 use backend defaults when empty; interval_ms is the polling interval.
/// 错误处理：resource_dir 获取失败时继续使用开发目录候选；注入器或 DLL 缺失由服务层返回明确错误。
/// Error handling: resource_dir failures fall back to development candidates; missing injector or DLL paths are reported by the service layer.
/// 中文关键词：进程监控，启动监控，注入监控，默认路径，资源目录
/// English keywords: process monitor, start monitor, inject monitor, default path, resource directory
#[tauri::command]
pub async fn start_process_watcher(
    watcher: tauri::State<'_, ProcessMonitorService>,
    app_handle: tauri::AppHandle,
    injector_x64: String,
    injector_x86: String,
    dll_x64: String,
    dll_x86: String,
    interval_ms: u32,
) -> Result<bool, String> {
    // ProcessMonitorService 需要可变引用，但 Tauri state 是共享引用
    // 使用内部 Mutex 来获取可变性
    let resource_dir = app_handle.path().resource_dir().ok();
    watcher.start_with_resource_dir(
        &injector_x64,
        &injector_x86,
        &dll_x64,
        &dll_x86,
        interval_ms,
        resource_dir.as_deref(),
    )
}

/// 函数名称：stop_process_watcher
/// 函数作用：停止进程监控服务。
/// Purpose: Stops the process monitor service.
/// 调用方：前端设置页面
/// Called by: Frontend settings page
/// 中文关键词：停止监控，停止进程监控
/// English keywords: stop monitor, stop process monitor
#[tauri::command]
pub async fn stop_process_watcher(
    watcher: tauri::State<'_, ProcessMonitorService>,
) -> Result<bool, String> {
    watcher.stop()?;
    Ok(true)
}

/// 函数名称：get_process_watcher_status
/// 函数作用：查询 APIHook 进程监控 watcher 是否正在运行。
/// Purpose: Queries whether the APIHook process monitor watcher is running.
/// 调用方：前端 getMonitoringRuntimeStatus。
/// Called by: frontend getMonitoringRuntimeStatus.
/// 被调用方：ProcessMonitorService::is_running。
/// Calls: ProcessMonitorService::is_running.
/// 返回值说明：true 表示 watcher 已启动，false 表示未启动。
/// Returns: true when the watcher is started, false when stopped.
/// 中文关键词：APIHook状态，进程监控状态，运行态查询
/// English keywords: APIHook status, process monitor status, runtime query
#[tauri::command]
pub async fn get_process_watcher_status(
    watcher: tauri::State<'_, ProcessMonitorService>,
) -> Result<bool, String> {
    Ok(watcher.is_running())
}

/// 函数名称：set_signed_process_list
/// 函数作用：设置已签名进程白名单，预填签名缓存。
/// Purpose: Sets the signed process whitelist, pre-fills signature cache.
/// 调用方：前端初始化时传递已知可信进程
/// Called by: Frontend on init passing known trusted processes
/// 中文关键词：签名白名单，可信进程列表
/// English keywords: signature whitelist, trusted process list
#[tauri::command]
pub async fn set_signed_process_list(
    watcher: tauri::State<'_, ProcessMonitorService>,
    paths: Vec<String>,
) -> Result<u32, String> {
    watcher.set_signed_list(&paths)
}

/// 函数名称：poll_new_pids
/// 函数作用：轮询获取新发现的进程 PID 列表。
/// Purpose: Polls for newly discovered process PID list.
/// 调用方：前端事件轮询
/// Called by: Frontend event polling
/// 中文关键词：轮询新进程，PID列表
/// English keywords: poll new processes, PID list
#[tauri::command]
pub async fn poll_new_pids(
    watcher: tauri::State<'_, ProcessMonitorService>,
) -> Result<Vec<u32>, String> {
    watcher.poll_new_pids()
}

File: src/api/process.ts
/**
 * 进程控制 API 封装
 * Process control API wrapper
 *
 * 封装进程暂停、恢复、终止、拦截和进程监控服务操作。
 * Wraps process suspend, resume, terminate, interception, and process watcher operations.
 */
import { invoke } from '@tauri-apps/api/core'
import { resolveResource } from '@tauri-apps/api/path'

/** 启动进程监控需要的原生文件路径/Native paths required to start process watcher */
export interface ProcessWatcherStartOptions {
  injectorX64: string
  injectorX86: string
  dllX64: string
  dllX86: string
  intervalMs: number
}

const DEFAULT_PROCESS_WATCHER_INTERVAL_MS = 2000

/**
 * 函数名称：defaultProcessWatcherStartOptions
 * 函数作用：生成空路径默认启动参数，让后端沿已有默认候选路径解析进程监控注入器和 DLL。
 * Purpose: Builds default start options with empty paths so the backend resolves watcher injector and DLL paths through its existing candidate paths.
 * 调用方：configStore.setProcessMonitoring。
 * Called by: configStore.setProcessMonitoring.
 * 返回值说明：返回 start_process_watcher 所需的强类型参数对象，四个路径为空字符串。
 * Returns: strongly typed options required by start_process_watcher with four empty paths.
 * 中文关键词：进程监控，默认参数，后端路径解析，运行态控制
 * English keywords: process watcher, default options, backend path resolution, runtime control
 */
export function defaultProcessWatcherStartOptions(): ProcessWatcherStartOptions {
  return {
    injectorX64: '',
    injectorX86: '',
    dllX64: '',
    dllX86: '',
    intervalMs: DEFAULT_PROCESS_WATCHER_INTERVAL_MS
  }
}

/**
 * 函数名称：createProcessWatcherStartOptions
 * 函数作用：创建显式路径启动参数；用于后续需要由前端指定原生文件位置的场景。
 * Purpose: Creates explicit path start options for future scenarios where the frontend must specify native file locations.
 * 调用方：当前未发现明确调用方。
 * Called by: No confirmed caller at present.
 * 参数说明：nativePaths 为四个原生文件路径；intervalMs 可覆盖默认轮询间隔。
 * Parameters: nativePaths contains four native file paths; intervalMs can override the default polling interval.
 * 返回值说明：返回 start_process_watcher 所需的强类型参数对象。
 * Returns: strongly typed options required by start_process_watcher.
 * 中文关键词：进程监控，显式路径，强类型参数
 * English keywords: process watcher, explicit path, strongly typed options
 */
export function createProcessWatcherStartOptions(
  nativePaths: Omit<ProcessWatcherStartOptions, 'intervalMs'>,
  intervalMs = DEFAULT_PROCESS_WATCHER_INTERVAL_MS
): ProcessWatcherStartOptions {
  return { ...nativePaths, intervalMs }
}

/**
 * 函数名称：resolveDefaultProcessWatcherStartOptions
 * 函数作用：从 Tauri 资源目录解析进程监控注入器和 file_hook DLL 的路径。
 * Purpose: Resolves process watcher injector and file_hook DLL paths from the Tauri resource directory.
 * 调用方：当前未发现明确调用方；保留给需要前端显式传路径的场景。
 * Called by: No confirmed caller at present; kept for scenarios requiring frontend-explicit paths.
 * 被调用方：@tauri-apps/api/path.resolveResource。
 * Calls: @tauri-apps/api/path.resolveResource.
 * 返回值说明：返回 start_process_watcher 所需的强类型参数对象。
 * Returns: strongly typed options required by start_process_watcher.
 * 错误处理：资源路径解析失败时向上抛出，由设置开关回滚 UI 状态。
 * Error handling: resource resolution failures are thrown so the settings toggle can roll back UI state.
 * 中文关键词：进程监控，原生资源路径，注入器，文件钩子，默认参数
 * English keywords: process watcher, native resource path, injector, file hook, default options
 */
export async function resolveDefaultProcessWatcherStartOptions(): Promise<ProcessWatcherStartOptions> {
  const [injectorX64, injectorX86, dllX64, dllX86] = await Promise.all([
    resolveResource('native/bin/win32-x64/file_hook_injector.exe'),
    resolveResource('native/bin/win32-x86/file_hook_injector.exe'),
    resolveResource('native/bin/win32-x64/file_hook_detours.dll'),
    resolveResource('native/bin/win32-x86/file_hook_detours.dll'),
  ])

  return {
    injectorX64,
    injectorX86,
    dllX64,
    dllX86,
    intervalMs: DEFAULT_PROCESS_WATCHER_INTERVAL_MS
  }
}

/**
 * 函数名称：suspendProcess
 * 函数作用：挂起指定 PID 的进程。
 * Purpose: Suspends the process identified by PID.
 */
export async function suspendProcess(pid: number): Promise<boolean> {
  return await invoke('suspend_process', { pid })
}

/**
 * 函数名称：resumeProcess
 * 函数作用：恢复指定 PID 的进程。
 * Purpose: Resumes the process identified by PID.
 */
export async function resumeProcess(pid: number): Promise<boolean> {
  return await invoke('resume_process', { pid })
}

/**
 * 函数名称：terminateProcess
 * 函数作用：终止指定 PID 的进程。
 * Purpose: Terminates the process identified by PID.
 */
export async function terminateProcess(pid: number): Promise<boolean> {
  return await invoke('terminate_process', { pid })
}

/**
 * 函数名称：handleInterception
 * 函数作用：处理用户拦截决策（放行/阻止）。
 * Purpose: Handles user interception decision (allow/block).
 */
export async function handleInterception(pid: number, action: 'allow' | 'block'): Promise<boolean> {
  return await invoke('handle_interception', { pid, action })
}

/**
 * 函数名称：getInterceptionQueue
 * 函数作用：获取当前拦截队列中的暂停进程 PID。
 * Purpose: Gets paused process PIDs in current interception queue.
 */
export async function getInterceptionQueue(): Promise<number[]> {
  return await invoke('get_interception_queue')
}

/**
 * 函数名称：clearInterceptionQueue
 * 函数作用：清空拦截队列。
 * Purpose: Clears the interception queue.
 */
export async function clearInterceptionQueue(): Promise<boolean> {
  return await invoke('clear_interception_queue')
}

/**
 * 函数名称：peekCurrentInterception
 * 函数作用：获取当前正在展示的拦截条目（若有），供拦截窗口初始化后主动拉取。
 * Purpose: Gets the currently shown interception entry (if any) for the interception window to pull after initialization.
 */
export async function peekCurrentInterception(): Promise<Record<string, unknown> | null> {
  return await invoke('peek_current_interception')
}

/**
 * 函数名称：startProcessWatcher
 * 函数作用：启动后端进程监控服务，轮询新进程并对需要监控的进程注入文件 Hook。
 * Purpose: Starts the backend process watcher, polling new processes and injecting file Hook where needed.
 * 调用方：configStore.setProcessMonitoring。
 * Called by: configStore.setProcessMonitoring.
 * 被调用方：Tauri invoke start_process_watcher。
 * Calls: Tauri invoke start_process_watcher.
 * 参数说明：options 包含 x64/x86 注入器、x64/x86 DLL 和轮询间隔。
 * Parameters: options contains x64/x86 injector paths, x64/x86 DLL paths, and polling interval.
 * 返回值说明：后端返回 true 表示命令执行成功。
 * Returns: backend returns true when command succeeds.
 * 中文关键词：启动进程监控，新进程轮询，注入器，运行态控制
 * English keywords: start process watcher, new process polling, injector, runtime control
 */
export async function startProcessWatcher(options: ProcessWatcherStartOptions): Promise<boolean> {
  return await invoke('start_process_watcher', {
    injectorX64: options.injectorX64,
    injectorX86: options.injectorX86,
    dllX64: options.dllX64,
    dllX86: options.dllX86,
    intervalMs: options.intervalMs
  })
}

/**
 * 函数名称：stopProcessWatcher
 * 函数作用：停止后端进程监控服务。
 * Purpose: Stops the backend process watcher service.
 * 调用方：configStore.setProcessMonitoring。
 * Called by: configStore.setProcessMonitoring.
 * 被调用方：Tauri invoke stop_process_watcher。
 * Calls: Tauri invoke stop_process_watcher.
 * 返回值说明：后端返回 true 表示命令执行成功。
 * Returns: backend returns true when command succeeds.
 * 中文关键词：停止进程监控，运行态控制
 * English keywords: stop process watcher, runtime control
 */
export async function stopProcessWatcher(): Promise<boolean> {
  return await invoke('stop_process_watcher')
}

/**
 * 函数名称：getProcessWatcherStatus
 * 函数作用：读取 APIHook 进程监控 watcher 是否正在运行。
 * Purpose: Reads whether the APIHook process watcher is running.
 * 调用方：getMonitoringRuntimeStatus。
 * Called by: getMonitoringRuntimeStatus.
 * 被调用方：Tauri invoke get_process_watcher_status。
 * Calls: Tauri invoke get_process_watcher_status.
 * 返回值说明：true 表示 APIHook watcher 已启动，false 表示未启动。
 * Returns: true when the APIHook watcher is started, false when stopped.
 * 中文关键词：APIHook状态，进程监控状态，运行态查询
 * English keywords: APIHook status, process watcher status, runtime query
 */
export async function getProcessWatcherStatus(): Promise<boolean> {
  return await invoke('get_process_watcher_status')
}

/**
 * 函数名称：pollNewPids
 * 函数作用：读取进程监控服务发现的新 PID 列表，后端读取后会清空队列。
 * Purpose: Reads newly discovered PIDs from the process watcher; backend drains the queue after reading.
 * 调用方：OverviewPage 或后续实时进程提示组件。
 * Called by: OverviewPage or future real-time process notification components.
 * 被调用方：Tauri invoke poll_new_pids。
 * Calls: Tauri invoke poll_new_pids.
 * 返回值说明：返回新发现的 PID 数组。
 * Returns: array of newly discovered PIDs.
 * 中文关键词：新PID，进程轮询，进程监控状态
 * English keywords: new PID, process polling, process watcher status
 */
export async function pollNewPids(): Promise<number[]> {
  return await invoke('poll_new_pids')
}
